sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java

```diff
@@ -14,82 +17,189 @@
+     * @return name for the transfer work
+     */
+    static String toTransferUniqueWorkName(long transferId) {
+        return "azure_transfer_" + transferId;
+    }
+
+    /**
+     * A builder to configure and build a {@link TransferClient}.
      */
-    public long upload(String containerName, String blobName, File file) {
-        BlobUploadEntity blob = new BlobUploadEntity(containerName, blobName, file);
-        List<BlockUploadEntity> blocks = BlockUploadEntity.createEntitiesForFile(file, Constants.DEFAULT_BLOCK_SIZE);
-        long uploadId = this.db.uploadDao().createUploadRecord(blob, blocks);
-        Log.v(TAG, "upload(): upload record created: " + uploadId);
-
-        StorageBlobClientsMap.put(uploadId, this.blobClient);
-        Constraints constraints = new Constraints.Builder()
-            .setRequiredNetworkType(NetworkType.CONNECTED)
-            .build();
-
-       Data inputData = new Data.Builder()
-            .putLong(UploadWorker.Constants.INPUT_BLOB_UPLOAD_ID_KEY, uploadId)
-            .build();
-
-        OneTimeWorkRequest uploadWorkRequest = new OneTimeWorkRequest
-            .Builder(UploadWorker.class)
-            .setConstraints(constraints)
-            .setInputData(inputData)
-            .build();
-        Log.v(TAG, "upload(): enqueuing UploadWorker: " + uploadId);
-        WorkManager.getInstance(this.context)
-            .beginUniqueWork("file_upload_" + uploadId, ExistingWorkPolicy.KEEP, uploadWorkRequest)
-            .enqueue();
-        return uploadId;
+    public static final class Builder {
+        // the application context.
+        private Context context;
+        // the default storage client for all transfers.
+        private StorageBlobClient storageBlobClient;
+        // the network type required for transfers.
+        private @TransferNetworkType int networkType;
+        // the executor for internal book keeping.
+        private SerialExecutor serialBackgroundExecutor;
+
+        /**
+         * Create a new {@link TransferClient} builder.
+         */
+        public Builder(@NonNull Context context) {
+            this.context = context;
+        }
+
+        /**
+         * Set the storage blob client to use for all transfers.
+         *
+         * @param storageBlobClient the storage blob client
+         * @return Builder with provided Storage Blob Client set
+         */
+        public Builder setStorageClient(@NonNull StorageBlobClient storageBlobClient) {
+            this.storageBlobClient = storageBlobClient;
+            return this;
+        }
+
+        /**
+         * Sets the particular {@link TransferNetworkType} the device should be in
+         * for running the transfers.
+         *
+         * @param networkType The type of network required for transfers to run
+         * @return Builder with provided network type set
+         */
+        public Builder setRequiredNetworkType(@TransferNetworkType int networkType) {
```

Comment:
I'm not opposed to taking `Constraints` instance directly. I think its good to honor additional constrains given those options make sense (like charging required, battery low).

Looks like there is no nice way to create a new `Constraints` instance from an existing one along with mutation (like `newBuilder()` on `Constraints`). Looking into it.

Reply:
It might be odd, but maybe we can just accept a `Constraints.Builder`? There's also a public `setRequiredNetworkType` method on the configuration but it has an annotation to hint to the IDE that it's not meant to be part of their public API. We could consider suppressing that lint (though, same concerns as before - this can be brittle).

Reply:
I think we should do some exploration around, whether taking builder as a param is a pattern in community. 

We can also think of taking `Constraints` instance and create a copy of it (by going through each property, bit verbose though) with required mutation applied.

Reply:
what do you think?

Reply:
I thought of just making a copy, but annoyingly there are not getters for every property that can be set by the builder. The `addContentUriTrigger` and all of the `setTrigger...` methods on the builder end up setting fields of the `Constraints` object which have no public getters :(

I can't think of any use case where a customer would want to add / configure such a trigger for their transfer but if they did we would miss it. Maybe the best approach is to just lift the other four `Constraints.Builder` methods we care about into our `TransferClient.Builder`:
- setRequiresBatteryNotLow
- setRequiresCharging
- setRequiresDeviceIdle
- setRequiresStorageNotLow

Also I think I'd be fine if we changed this method to accept an `androidx.work.NetworkType`, I don't think I have a concern about leaking this type from WorkManager, we can assume customers are already reasonably familiar with this type or at least can look up the docs for it from Google. But I could be convinced either way, I just don't see a ton of value being provided by wrapping it with our own enum.

Reply:
hey @bsiegel thank you for doing all these investigations and adding lot of values while I'm heads down into immediate customer issue.

Except `setRequiredNetworkType`, all other setters you listed takes primitive boolean so only type we have to consider is `androidx.work.NetworkType`. I don't have any concern in exposing `androidx.work.NetworkType` directly. Let's go with that and during user study we can watch how users adapt to it and can adjust based on feedback, if any.

If app provide `androidx.work.NetworkType.NOT_REQUIRED` then I think we can just throw an IllegalArgumentException.

Reply:
I'm onboard with that approach 👍